lib/smooth/application/io/i2c/DS3231.cpp

```diff
@@ -0,0 +1,658 @@
+                            / (BCD2Uint8_t(data[5] & 0x7f) - 1) / BCD2Uint8_t(data[4]);
+
+                return sys_days(date)
+                       + hours(
+                    (data[2] & 0x40)
+                        ? BCD2Uint8_t(data[2] & 0x1F) - 1 + (data[2] & 0x20) ? 12 : 0 /* 12H */
+                        : BCD2Uint8_t(data[2] & 0x3F) /* 24H */
+                    )
+                       + minutes(BCD2Uint8_t(data[1]))
+                       + seconds(BCD2Uint8_t(data[0]));
+            }
+
+            bool DS3231::SetDateTime(system_clock::time_point& dateTime)
+            {
+                // TODO check that dateTime is valid
+                auto timePointWithDaysResolution = floor<days>(dateTime);
+                auto date = year_month_day(timePointWithDaysResolution);
+                auto time = make_time(dateTime - timePointWithDaysResolution);
+
+                std::vector<uint8_t> data;
+                data.push_back(Register::DateTime);
+                data.push_back(Uint8_t2BCD(static_cast<uint8_t>(time.seconds().count())));
+                data.push_back(Uint8_t2BCD(static_cast<uint8_t>(time.minutes().count())));
+                data.push_back(Uint8_t2BCD(static_cast<uint8_t>(time.hours().count())));
+                data.push_back(Uint8_t2BCD(static_cast<uint8_t>((year_month_weekday(date).weekday_indexed().weekday()
+                                                                 - Sunday).count())));
+                data.push_back(Uint8_t2BCD( static_cast<uint8_t>(unsigned(date.day()))));
+                uint8_t year = static_cast<uint8_t>((int)date.year() - 2000);
+                data.push_back(Uint8_t2BCD(
+                    static_cast<uint8_t>((unsigned(date.month()) + 1) | (year >= 100 ? 0x80 : 0))));
+                data.push_back(Uint8_t2BCD(static_cast<uint8_t>(year - (year >= 100 ? 100 : 0))));
+
+                return write(address, data, true);
+            }
+
+            bool DS3231::SetAlarmOne(AlarmOneRate rate, system_clock::time_point const& timePoint)
+            {
+                // TODO check that dateTime is valid
+                auto timePointWithDaysResolution = floor<days>(timePoint);
+                auto date = year_month_day(timePointWithDaysResolution);
+                auto time = make_time(timePoint - timePointWithDaysResolution);
+
+                std::vector<uint8_t> data;
+                data.push_back(Register::Alarm1);
+                data.push_back(rate >= SecondsMatch ? Uint8_t2BCD(static_cast<uint8_t>(time.seconds().count())) : 0x80);
+                data.push_back(rate >= MinutesAndSecondsMatch ? Uint8_t2BCD(
+                    static_cast<uint8_t>(time.minutes().count())) : 0x80);
+                data.push_back(rate >=
+                HoursMinutesAndSecondsMatch ? Uint8_t2BCD(static_cast<uint8_t>(time.hours().count())) : 0x80);
+                data.push_back(
+                    rate >= DateHoursMinutesAndSecondsMatch
```

Comment:
Can we avoid nested ?: operators? It makes it somewhat hard to follow, especially with the long statements.